File: cmd/ripperx/web/i18n.js
'use strict';

// Every word these pages show comes out of a locale file rather than out of
// the markup or the code, so adding a language means adding one file and
// changing nothing else. The files live beside the pages and are served as
// they are; the list of them and which one this server prefers come from
// /api/locales, which is readable without signing in because the sign-in
// page needs its own words too.
//
// Three dictionaries are consulted in turn: the one the reader chose, the
// one this server was configured with, and English. A string missing from a
// half-finished translation therefore appears in the server's language, or
// failing that in English, rather than as a gap - which is what makes it
// possible to add a language a few lines at a time.

const i18n = (() => {
  const builtinFallback = 'en';
  const stored = 'ripperx.lang';

  const state = {
    code: builtinFallback,
    serverDefault: builtinFallback,
    languages: [{ code: builtinFallback, name: 'English' }],
    // The three dictionaries, most specific first.
    chain: [],
  };

  // Plural rules, by the name a locale file declares. English wants two
  // forms and Russian three, and a language that says nothing gets the
  // two-form rule - which is wrong for some languages and is at least
  // wrong in a way its translator can fix by naming a rule here.
  const pluralRules = {
    'one-other': (n) => (n === 1 ? 'one' : 'other'),
    russian: (n) => {
      const mod10 = n % 10, mod100 = n % 100;
      if (mod10 === 1 && mod100 !== 11) return 'one';
      if (mod10 >= 2 && mod10 <= 4 && (mod100 < 12 || mod100 > 14)) return 'few';
      return 'many';
    },
  };

  function lookup(key) {
    for (const dict of state.chain) {
      const v = dict[key];
      if (v !== undefined) return v;
    }
    // A key with no string anywhere shows as itself. It is not a nice thing
    // to read, and that is the point: it is a bug, and it should look like
    // one rather than like an empty label.
    return key;
  }

  // fill substitutes {name} placeholders. A name with no value is left
  // alone, so a mistake in a translation shows what was expected.
  function fill(s, vars) {
    if (!vars) return s;
    return String(s).replace(/\{(\w+)\}/g, (whole, name) =>
      vars[name] === undefined || vars[name] === null ? whole : String(vars[name]));
  }

  function t(key, vars) { return fill(lookup(key), vars); }

  // tn picks the form that goes with a count and passes the count in as
  // {n}, so a translation writes "{n} discs" without the caller knowing how
  // many forms the language has.
  function tn(key, n, vars) {
    const rule = pluralRules[lookup('_plural')] || pluralRules['one-other'];
    const form = rule(Math.abs(Number(n) || 0));
    let s = lookup(`${key}.${form}`);
    if (s === `${key}.${form}`) s = lookup(`${key}.other`);
    return fill(s, Object.assign({ n }, vars));
  }

  // apply writes every string into the markup that asked for one. The
  // attributes are separate because a placeholder and a title are not text
  // nodes and setting them as one would put the words in the wrong place.
  function apply(root) {
    const scope = root || document;
    const put = (attr, set) => {
      for (const node of scope.querySelectorAll(`[${attr}]`)) {
        set(node, t(node.getAttribute(attr)));
      }
    };
    put('data-i18n', (n, s) => { if (n.textContent !== s) n.textContent = s; });
    put('data-i18n-placeholder', (n, s) => { n.placeholder = s; });
    put('data-i18n-title', (n, s) => { n.title = s; });
    put('data-i18n-label', (n, s) => { n.setAttribute('aria-label', s); });
  }

  async function fetchLocale(code) {
    try {
      const r = await fetch(`/locales/${encodeURIComponent(code)}.json`, { cache: 'no-cache' });
      if (!r.ok) return null;
      return await r.json();
    } catch (e) {
      return null;
    }
  }

  // chosen is the reader's own preference: what the URL says, then what
  // this browser remembered, then what the server was configured with. The
  // URL comes first so a link can be sent in a particular language.
  function chosen() {
    const fromURL = new URLSearchParams(window.location.search).get('lang');
    if (fromURL) return fromURL;
    try {
      const saved = window.localStorage.getItem(stored);
      if (saved) return saved;
    } catch (e) { /* a browser with storage turned off still gets a language */ }
    return state.serverDefault;
  }

  function known(code) {
    return state.languages.some((l) => l.code === code);
  }

  async function start() {
    try {
      const r = await fetch('/api/locales', { cache: 'no-cache' });
      if (r.ok) {
        const cat = await r.json();
        if (cat.languages && cat.languages.length) state.languages = cat.languages;
        if (cat.default) state.serverDefault = cat.default;
      }
    } catch (e) { /* the built-in defaults are enough to render English */ }

    let code = chosen();
    if (!known(code)) code = state.serverDefault;
    if (!known(code)) code = builtinFallback;
    state.code = code;

    // Loaded most specific first, and each only once: a server whose
    // default is English asks for one file, not three.
    const wanted = [];
    for (const c of [code, state.serverDefault, builtinFallback]) {
      if (c && !wanted.includes(c)) wanted.push(c);
    }
    const dicts = await Promise.all(wanted.map(fetchLocale));
    state.chain = dicts.filter((d) => d);
    document.documentElement.lang = code;
    apply(document);
  }

  // choose switches language. The page is reloaded rather than re-rendered:
  // the strings are woven through every list, every dialog and every job
  // line on screen, and reloading is both simpler and provably complete.
  function choose(code) {
    if (code === state.code) return;
    try { window.localStorage.setItem(stored, code); } catch (e) { /* not fatal */ }
    const url = new URL(window.location.href);
    url.searchParams.delete('lang');
    window.location.replace(url.toString());
  }

  return {
    t, tn, apply, start, choose,
    get code() { return state.code; },
    get languages() { return state.languages; },
  };
})();

const t = (key, vars) => i18n.t(key, vars);
const tn = (key, n, vars) => i18n.tn(key, n, vars);

File: cmd/ripperx/i18n.go
package main

import (
	"encoding/json"
	"fmt"
	"io/fs"
	"net/http"
	"path"
	"sort"
	"strings"
)

// The pages carry no English of their own: every word they show is looked up
// in a locale file, and the files are ordinary JSON served from the same
// place as the pages. Adding a language is therefore adding one file to
// web/locales - nothing here and nothing in the pages has to know about it,
// because this reads the directory rather than a list somebody maintains.
//
// Each file names itself in its own language, under "_name", so the menu is
// built from the files too.

// defaultLanguage is the language a server serves when it was not told
// otherwise, and the last fallback for a string missing everywhere else.
const defaultLanguage = "en"

type language struct {
	Code string `json:"code" doc:"the file's name, and what to pass as the lang parameter"`
	Name string `json:"name" doc:"what the language calls itself"`
}

type localesResponse struct {
	Default   string     `json:"default" doc:"the language this server was configured with; the fallback for anything missing from another"`
	Languages []language `json:"languages" doc:"every language installed, by the name it calls itself"`
}

// readLanguages lists the locale files that came with this binary. A file
// that is not readable JSON is left out rather than fatal: a broken
// translation should cost that language, not the server.
func readLanguages(web fs.FS) []language {
	entries, err := fs.ReadDir(web, "locales")
	if err != nil {
		return nil
	}
	var out []language
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		code := strings.TrimSuffix(e.Name(), ".json")
		b, err := fs.ReadFile(web, path.Join("locales", e.Name()))
		if err != nil {
			continue
		}
		var strs map[string]string
		if json.Unmarshal(b, &strs) != nil {
			continue
		}
		name := strs["_name"]
		if name == "" {
			name = code
		}
		out = append(out, language{Code: code, Name: name})
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Code < out[j].Code })
	return out
}

// checkLanguage refuses to start on a language that is not installed. A
// server silently serving English because of a typo in its settings file is
// worse than one that will not start and says which languages it has.
func checkLanguage(code string, have []language) error {
	for _, l := range have {
		if l.Code == code {
			return nil
		}
	}
	names := make([]string, 0, len(have))
	for _, l := range have {
		names = append(names, l.Code)
	}
	return fmt.Errorf("lang: %q is not one of the languages this build carries (%s)",
		code, strings.Join(names, ", "))
}

func (s *server) handleLocales(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, localesResponse{Default: s.lang, Languages: s.languages})
}
